R/logger.R

```diff
@@ -67,11 +67,50 @@ fallback_namespace <- function(namespace) {
     namespace
 }
 
+#' Base Logging Function
+#' @param fun_name string a full name of log function
+#' @param arg see \code{\link{log_levels}}
+#' @param namespace logger namespace
+#' @param index index of the logger within the namespace
+#' @return currently set or return log function property
+#' @keywords internal
+log_base <- function(fun_name, arg, namespace, index) {
+
+    if (length(namespace) > 1) {
+        for (ns in namespace) {
+            log_base(fun_name, arg, ns, index)
+        }
+        return(invisible())
+    }
 
-## TODO DRY the below 4 functions
+    fun_name_base <- strsplit(fun_name, "_")[[1]][2]
+
+    configs <- get(fallback_namespace(namespace), envir = namespaces)
+    config  <- configs[[min(index, length(configs))]]
+
+    if (fun_name_base == "threshold") {
+      if (is.null(arg)) {
+        return(config[[fun_name_base]])
+      }
+      config[[fun_name_base]] <- validate_log_level(arg)
+    } else {
+      if (is.null(arg)) {
+        res <- config[[fun_name_base]]
+        if (!is.null(attr(res, "generator"))) {
+          res <- parse(text = attr(res, "generator"))[[1]]
+        }
+        return(res)
+      }
+
+      config[[fun_name_base]] <- arg
+    }
+
+    configs[[min(index, length(config) + 1)]] <- config
+    assign(namespace, configs, envir = namespaces)
+}
 
 #' Get or set log level threshold
-#' @param level see \code{\link{log_levels}}
+#' @param level see \code{\link{log_levels}}, default NULL
 #' @param namespace logger namespace
```

Comment:
`namespace` and `index` can `inheritParams` from the base setter

Reply:
reverted